HQ support for the Reminders App

kaapstorm opened the pull request:
## Summary

Jira: [SC-953](https://dimagi-dev.atlassian.net/browse/SC-953)

This PR implements the HQ side of the work to support the Reminders App.

Case List and Form settings include a text field under "Settings" > "Advanced" labeled "Session Endpoint ID". The value is used to generate an `<endpoint>` node in the app's **suite.xml** file.

This PR includes a couple of commented-out "**TO DO**"s to hide/disable the new functionality if the app is using a version of CommCare that does not support session endpoints. I am keen to hear feedback on whether people think this is useful to uncomment once the upcoming version of CommCare is released, or whether I should drop the commented-out code.

:fish: :blowfish: :tropical_fish: :octopus: Easier to review by commit.

fyi @shubham1g5 


## Feature Flag

'Enable session endpoints'


## Product Description

HQ support for the Reminders App. 


## Safety Assurance

- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I am certain that this PR will not introduce a regression for the reasons below


### Automated test coverage

PR includes automated test coverage. If you feel coverage is insufficient, please let me know.


### QA Plan

This branch is currently deployed to staging for review.


### Safety story

* Unit test coverage
* Tested locally
* Tested on Staging


### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

kaapstorm committed:
SessionEndpoint XML model

kaapstorm committed:
SessionEndpointContributor class

kaapstorm committed:
Add session_endpoint_ids to form and module models

kaapstorm committed:
Add toggle and supports_session_endpoints()

kaapstorm committed:
Add session endpoints to suite.xml

kaapstorm committed:
Add session endpoint ID to UI

kaapstorm committed:
⚖ Test

proteusvacuum commented on corehq/toggles.py:
```diff
@@ -1544,6 +1544,16 @@ def _commtrackify(domain_name, toggle_is_enabled):
+SESSION_ENDPOINTS = StaticToggle(
+    'session_endpoints',
+    'Enable session endpoints',
+
```
nit: extra newline here

proteusvacuum commented on corehq/apps/app_manager/suite_xml/sections/endpoints.py:
```diff
@@ -0,0 +1,72 @@
+    for endpoint_id in endpoint_ids:
+        frame.add_datum(StackDatum(
+            id=endpoint_id,
+            value=session_var(endpoint_id))
```
I'm maybe not completely understanding how this works, but how is this session variable populated in the reminders app workflow? Doesn't that mean that a user will have had to already navigate to the case they want in CommCare before navigating the reminders app? Or is there some other way this is populated? 

The spec seems to suggest using an `<argument>` whose name is the value here, which I assume is something that the reminders app would pass in when this callout is triggered. 
```
<argument id=”person_case_id”/>
...
<datum id=”case_id” value=”$person_case_id”>
````

esoergel replied:
that sounds in line with Shubham's comment [here](https://dimagi-dev.atlassian.net/browse/SC-953?focusedCommentId=115731)

kaapstorm replied:
Yeah. This code is wrong. Shubham's comment is spot on, and clarifies the problem, and how to fix it. We need to have an `<argument>` for each case the form needs.

esoergel committed:
PR feedback - 2.51 is now on prod and staging

esoergel commented on corehq/toggles.py:
```diff
@@ -1563,6 +1563,16 @@ def _commtrackify(domain_name, toggle_is_enabled):
+    'session_endpoints',
+    'Enable session endpoints',
+
+    TAG_SOLUTIONS,
```
Know if this should be `TAG_SOLUTIONS_LIMITED`, `OPEN`, or `CONDITIONAL`?

kaapstorm replied:
I'm not sure. But I bet @karmuno knows.

esoergel replied:
Tagged it "limited use" for now - it's an easy thing to change later as desired.

esoergel committed:
This text appears on both module and form settings pages

esoergel commented on corehq/apps/app_manager/templates/app_manager/partials/session_endpoints.html:
```diff
@@ -0,0 +1,26 @@
+                   {% endcomment %}
+                   class="{% css_field_class %}">
+                <input class="code form-control"
+                       type="text"
```
Know if this needs to be a valid slug or anything?  Should we slap some html5 validation on it?

kaapstorm replied:
A slug sounds right.

orangejenny commented on corehq/apps/app_manager/tests/test_session_endpoints.py:
```diff
@@ -0,0 +1,55 @@
+from unittest.mock import patch
```
Minor naming convention: I'd call this file `test_suite_session_endpoints`.

orangejenny commented on corehq/apps/app_manager/feature_support.py:
```diff
@@ -177,3 +177,10 @@ def enable_markdown_tables(self):
+    def supports_session_endpoints(self):
+        return (
+            toggles.SESSION_ENDPOINTS.enabled(self.domain)
+            # and self._require_minimum_version('2.51')  # TODO: Uncomment after release
```
Also if 2.51 isn't created, you can create it and mark it superuser-only.

esoergel replied:
2.51 is out on staging and prod now, so I uncommented this and a similar one in the form template

esoergel commented on corehq/apps/app_manager/suite_xml/generator.py:
```diff
@@ -81,6 +86,7 @@ def generate_suite(self):
         entries = EntriesContributor(self.suite, self.app, self.modules, self.build_profile_id)
         menus = MenuContributor(self.suite, self.app, self.modules, self.build_profile_id)
         remote_requests = RemoteRequestContributor(self.suite, self.app, self.modules)
+        session_endpoints = SessionEndpointContributor(self.suite, self.app, self.modules)
```
@orangejenny do you have strong opinions about this?

orangejenny replied:
I'd favor consistency and keep the code as it is.

esoergel replied:
Fair enough.

kaapstorm commented on corehq/apps/app_manager/templates/app_manager/partials/session_endpoints.html:
```diff
@@ -0,0 +1,26 @@
+                       type="text"
+                       id="session_endpoint_id"
+                       name="session_endpoint_id"
+                       {# CommCare currently only supports one session endpoint #}
```
This might be worth confirming with @shubham1g5 My understanding is that we want (wanted?) to be able to offer more than one in the future. (?)

shubham1g5 replied:
We will only have one session endpoint per form/module. Though we definitely want to have multiple session endpoints overall for different screens of an app

esoergel replied:
Thanks - I modified the code to reflect this.

esoergel committed:
Operate with single endpoint ID per endpoint

Mobile only supports one, and I don't think we have a current use-case
for multiple

esoergel committed:
Validate that session endpoint ID is slug

esoergel committed:
Rejigger tests, make db-free

esoergel committed:
Prove that users can remove session endpoints

sravfeyn committed:
Merge master

esoergel committed:
Redefine session endpoints in suite file

esoergel committed:
Extract shared validation logic

esoergel commented on corehq/apps/app_manager/tests/test_suite_session_endpoints.py:
```diff
@@ -0,0 +1,91 @@
+@patch_validate_xform()
+@patch_get_xform_resource_overrides()
+@flag_enabled('SESSION_ENDPOINTS')
+class SessionEndpointTests(SimpleTestCase, TestXmlMixin):
```
@shubham1g5 this test suite contains a few examples of generated suite files - a registration form, followup form, and a case list.  Do you have any more complex examples that would be useful to add?  Perhaps weirder scenarios like child case lists, for example.  I'm not confident all edge-cases are handled appropriately.

shubham1g5 replied:
This looks great, I agree that it would be nice to add an example for parent child workflow. [This form](https://staging.commcarehq.org/a/another-upstream/apps/view/b1e763d627164bccb1359585e0a5b7f7/form/05e6f3afe12c4048a7b26bd46963fd66/source/#form/the_show_is_in_case_name) in particular looks like a good example to cover. The suite file for it defines the form entry with the datums - 

````
 <session>
      <datum id="parent_id" nodeset="instance('casedb')/casedb/case[@case_type='song'][@status='open']" value="./@case_id" detail-select="m2_case_short"/>
      <datum id="case_id" nodeset="instance('casedb')/casedb/case[@case_type='show'][@status='open'][index/parent=instance('commcaresession')/session/data/parent_id]" value="./@case_id" detail-select="m3_case_short" detail-confirm="m3_case_long"/>
    </session>
````

so the session endpoint should define 2 arguments for this aka `$parent_id` and `$case_id`

esoergel replied:
Great, just added a couple of examples for that.   Let me know if anything else comes up that should be tested explicitly.

esoergel committed:
Ensure no duplicate endpoint IDs are created

kaapstorm commented on corehq/apps/app_manager/suite_xml/sections/endpoints.py:
```diff
@@ -0,0 +1,69 @@
+    def _workflow_helper(self):
+        return WorkflowHelper(self.suite, self.app, self.modules)
+
+    def _make_session_endpoint(self, id_string, module, endpoint_id):
```
<3

kaapstorm reviewed (commented):
I can't approve this PR because I opened it. But please consider it approved. This is looking great.

esoergel committed:
Test child case list and form

esoergel reviewed (approved):
AFAIK this PR is now ready to go.  I'm gonna hold off on merge to give @shubham1g5 time to try it out on staging, but any reviewers waiting for stability can now take another pass.

esoergel committed:
Don't copy session_endpoint_id to shadows

proteusvacuum commented on corehq/apps/app_manager/suite_xml/sections/endpoints.py:
```diff
@@ -0,0 +1,69 @@
+            if isinstance(child, WorkflowDatumMeta):
+                arguments.append(Argument(id=child.id))
+                frame.add_datum(
+                    StackDatum(id=child.id, value=f"${child.id}")
```
Is it always obvious to the user what this id will be? I'm assuming for advanced modules that this will be something like `case_id_new_{case type}_{action index}`. Is that correct, or have I misunderstood what this is actually doing?

esoergel replied:
I'm actually not sure what happens on the other side of this equation.  The reminders app (or other future usage of this feature) _will_ need to know how to fill out this information, but I'm not certain how they do that.  It should be the same as how these data are made available in the session.  Do you have any thoughts around how to make it clearer?

proteusvacuum replied:
Got it. I guess if you are using advanced modules you are on your own anyway, and the format for the case ids are pretty well documented. I had been under the impression that the argument ID was something that the user would configure (so they would remain static if anything else about the app changes), but sounds like this happens automatically. I don't understand this feature well enough to know if that is an issue or not, so can leave it up to you.

esoergel replied:
Ah, gotcha - I think you're thinking of the endpoint ID.  That's an attribute of the `<endpoint />` node wrapping this.  In an earlier version of this PR, that was also used as the argument ID, but Shubham pointed out the argument ID should match the datum ID.  Honestly I don't have my head wrapped around exactly how that's used, but my understanding is the argument ID is to be used by the other app to figure out how to fill out the stack.

esoergel committed:
Don't cache workflow helper between endpoints

This resulted in only populating one of the endpoints' data

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into nh/reminders
